### PR TITLE
API improvements

### PR DESCRIPTION
**Pull request by UladBohdan**
* `imageUrl` is added to `ArticleBrand` and `ArticleCollection`. In addition, `ArticleBrand` has `imageUrlSmall` field.
* `firstName`, `lastName` and `bio` are Strings for regular users and locale maps for author users.
* `displayName` is returned for every User/Author. Equals to `firstName` if `lastName` is not provided and equals to `firstName[space]lastName` if both fields are provided. This is also localized for Authors.
* you can now `GET`, `PUT` and `DELETE` Article either by ID or by slug (of any localization). Covered with tests.
* creating `Article` TOGETHER with localizations in ONE API call (an update for POST `/api/articles` method).
* if no Diaries exist http error **204** is returned
* new response format for getOneDiary: `{ data, prev, next }`, where `prev` and `next` contain `month` and `day` fields indicating the closest diaries to switch to. `data` either contains a requested Diary or is empty.
* awesome Create and Update methods for articles! You can create/update with only call with localization nested in the request.
* a shit-ton of tests. But we might need more!

### CODE REVIEW

**Comment by UladBohdan** [comment]
[to return displayName for users](https://trello.com/c/iHWbpyLK/166-to-return-displayname-for-users)

File: src/api/user/model.js [hunk 1]
```diff
@@ -32,8 +32,9 @@ const UserSchema = new Schema({
   imageUrl: String,
 });
 
-UserSchema.virtual('name').get(function get() {
-  return `${this.firstName} ${this.lastName}`;
+UserSchema.virtual('displayName').get(function get() {
+  const postfix = this.lastName ? ` ${this.lastName}` : '';
+  return `${this.firstName}${postfix}`;
```
Comment (Drapegnik) [inline src/api/user/model.js]: `${this.firstName}${this.lastName || ''}`;

Comment (UladBohdan) [reply src/api/user/model.js]: That does not equal to what is written. I need a space before `lastName` if it is presented.

**Comment by UladBohdan** [comment]
[imageUrl for Brands and Collections](https://trello.com/c/I1W7Bsq3/168-imageurl-for-brands-and-collections)

**Comment by UladBohdan** [comment]
[User (not the author): replace localized map with strings!](https://trello.com/c/RaQEBiZC/164-user-not-the-author-replace-localized-map-with-strings)

**Comment by Drapegnik** [comment]
> getOneDiary: -> `{ data, prev, next }`

what if `next / prev` would be links to closes existing dairies?

File: src/api/article/controller.js [hunk 2]
```diff
@@ -45,10 +47,11 @@ export const getAll = ({ query, user }, res, next) => {
     .catch(next);
 };
 
-export const getOne = ({ params: { slug }, user }, res, next) =>
-  LocalizedArticle.findOne({ slug })
+export const getOne = ({ params: { slugOrId }, user }, res, next) =>
```
Comment (Drapegnik) [inline src/api/article/controller.js]: can we get query by virtual fields?

File: src/api/article/controller.js [hunk 3]
```diff
@@ -61,67 +64,173 @@ export const getOne = ({ params: { slug }, user }, res, next) =>
...
-    }
-    sendJson(res, code)(data);
-  } catch (err) {
-    next(err);
-  }
-};
+export const create = ({ body }, res, next) =>
+  // TODO(uladbohdan): to deprecate body.brand.
+  ArticleBrand.findOne({ slug: body.brand || body.brandSlug })
+    .then(checkIsFound) // Brand is required.
```
Comment (Drapegnik) [inline src/api/article/controller.js]: it return `404`, but should `400 + Brand is required.`

File: src/api/article/controller.js [hunk 4]
```diff
@@ -61,67 +64,173 @@ export const getOne = ({ params: { slug }, user }, res, next) =>
...
-    } catch (err) {
-      code = 400;
-      data = err;
-    }
-    sendJson(res, code)(data);
-  } catch (err) {
-    next(err);
-  }
-};
+export const create = ({ body }, res, next) =>
```
Comment (Drapegnik) [inline src/api/article/controller.js]: you have a [**promise hell**](https://habrahabr.ru/post/319094/) in this method, should be refactored with `async / await`

File: src/api/article/controller.js [hunk 5]
```diff
@@ -61,67 +64,173 @@ export const getOne = ({ params: { slug }, user }, res, next) =>
...
+                ...omit(body, ['locales']),
+                author: authorId,
+                brand: brandId,
+                collectionId,
+              })
+                .save()
+                .then(async article => {
+                  // Processing with localizations (Bundled API).
+                  if (body.locales) {
+                    await Promise.all(
```
Comment (Drapegnik) [inline src/api/article/controller.js]: const article.locales

File: src/api/article/controller.js [hunk 6]
```diff
@@ -61,67 +64,173 @@ export const getOne = ({ params: { slug }, user }, res, next) =>
...
+                          .then(({ _id: locId }) => article.locales.push(locId))
+                          .catch(next)
+                      )
+                    );
+                  }
+                  return article;
+                })
+                .then(article => article.save())
+                .then(({ _id: articleId }) =>
+                  Article.findOne({ _id: articleId })
```
Comment (Drapegnik) [inline src/api/article/controller.js]: imho can be extracted in some method, and used both in `getOne()` and here

File: src/api/article/controller.js [hunk 7]
```diff
@@ -61,67 +64,173 @@ export const getOne = ({ params: { slug }, user }, res, next) =>
...
+                    .then(sendJson(res))
+                    .catch(next)
+                )
+                .catch(next)
+            )
+            .catch(next)
+        )
+        .catch(next)
+    )
+    .catch(next);
```
Comment (Drapegnik) [inline src/api/article/controller.js]: download

🤣

File: src/api/article/controller.js [hunk 8]
```diff
@@ -61,67 +64,173 @@ export const getOne = ({ params: { slug }, user }, res, next) =>
...
 
-export const update = ({ params: { slug }, body }, res, next) =>
-  LocalizedArticle.findOne({ slug })
+export const update = ({ params: { slugOrId }, body }, res, next) =>
+  LocalizedArticle.findOne({ slug: slugOrId, active: true })
+    .then(result => (result && result.articleId) || (isValidId(slugOrId) && slugOrId))
```
Comment (Drapegnik) [inline src/api/article/controller.js]: lines above can extracted & reused

File: src/api/article/controller.js [hunk 9]
```diff
@@ -61,67 +64,173 @@ export const getOne = ({ params: { slug }, user }, res, next) =>
...
     .then(checkIsFound)
-    .then(({ articleId }) =>
-      Article.findOneAndUpdate({ _id: articleId }, body, { new: true })
-        .populate('author', POPULATE_OPTIONS.author)
-        .populate('brand', POPULATE_OPTIONS.brand)
-        .populate('collectionId', POPULATE_OPTIONS.collection)
-        .populate('locales', POPULATE_OPTIONS.locales)
```
Comment (Drapegnik) [inline src/api/article/controller.js]: it was such a small & beautiful method 😭

File: src/api/article/controller.js [hunk 10]
```diff
@@ -61,67 +64,173 @@ export const getOne = ({ params: { slug }, user }, res, next) =>
...
+                    defaultFields.brand = newBrand._id;
+                  }
+                  if (newCollection) {
+                    defaultFields.collectionId = newCollection._id;
+                  }
+                  if (newAuthor) {
+                    defaultFields.author = newAuthor._id;
+                  }
+
+                  return defaultFields;
```
Comment (Drapegnik) [inline src/api/article/controller.js]: what about case when need to remove author / collection?

File: src/api/article/controller.js [hunk 11]
```diff
@@ -61,67 +64,173 @@ export const getOne = ({ params: { slug }, user }, res, next) =>
...
+                })
+                .then(updateFields =>
+                  Article.findOneAndUpdate({ _id: articleId }, updateFields, { new: true })
+                    .then(checkIsFound)
```
Comment (Drapegnik) [inline src/api/article/controller.js]: ?

File: src/api/article/controller.js [hunk 12]
```diff
@@ -61,67 +64,173 @@ export const getOne = ({ params: { slug }, user }, res, next) =>
...
+                    .then(async article => {
+                      // Here Article is properly updated. Proceeding with locales.
+                      let articleOldLocales = [];
+                      if (body.locales) {
+                        articleOldLocales = article.locales.slice(); // Copying.
+                        article.locales.splice(0, article.locales.length); // Clearing with no reassignment.
```
Comment (Drapegnik) [inline src/api/article/controller.js]: ??

File: src/api/article/controller.js [hunk 13]
```diff
@@ -61,67 +64,173 @@ export const getOne = ({ params: { slug }, user }, res, next) =>
...
+                              .catch(next)
+                          )
+                        );
+                      }
+                      return { article, articleOldLocales };
+                    })
+                    .then(({ article, articleOldLocales }) =>
+                      article.save().then(async () => {
+                        // Here I want to remove unnecessary locales.
+                        // All locales which are in articleOldLocales and are not in article.locales
```
Comment (Drapegnik) [inline src/api/article/controller.js]:
```js
const toRemove = articleOldLocales.filter(l => !article.locales.includes(s))
```

File: src/api/article/localized/controller.js [hunk 14]
```diff
@@ -6,6 +6,7 @@ import Article from 'api/article/article.model';
 
 import LocalizedArticle from './model';
 
+// TODO(uladbohdan): to refactor using then-chaining.
```
Comment (Drapegnik) [inline src/api/article/localized/controller.js]: why?

File: src/api/specials/diary/controller.js [hunk 15]
```diff
@@ -1,12 +1,36 @@
...
+      const nextDate = new Date(date);
+      nextDate.setDate(date.getDate() + 1);
+
+      const [prevExists, nextExists] = await Promise.all(
+        [prevDate, nextDate].map(curDate =>
+          Diary.findOne({
+            locale,
+            colloquialDate: dateFormat(curDate, 'mm-dd'),
+            active: true,
+          }).then(d => Boolean(d))
```
Comment (Drapegnik) [inline src/api/specials/diary/controller.js]: just `.then(Boolean)`

File: src/api/user/model.js [hunk 16]
```diff
@@ -32,8 +38,20 @@ const UserSchema = new Schema({
   imageUrl: String,
 });
 
-UserSchema.virtual('name').get(function get() {
-  return `${this.firstName} ${this.lastName}`;
+UserSchema.virtual('displayName').get(function get() {
+  const joinNames = (firstName, lastName) => {
```
Comment (Drapegnik) [inline src/api/user/model.js]: extract to `utils`

File: src/api/user/model.js [hunk 17]
```diff
@@ -32,8 +38,20 @@ const UserSchema = new Schema({
...
+    const postfix = lastName ? ` ${lastName}` : '';
+    return `${firstName}${postfix}`;
+  };
+  if (this.role === 'author') {
+    const result = {};
+    Object.keys(this.firstName).forEach(locale => {
```
Comment (Drapegnik) [inline src/api/user/model.js]: in one line :)
```js
const result = _.fromPairs(
	Object
		.entries(this.firstName)
		.map(
			([l, name]) => [l, joinNames(name, this.lastName && this.lastName[l])]
		)
);

File: src/utils/validation.js [hunk 18]
```diff
@@ -1,8 +1,64 @@
...
+
+// These are the drafts for validators.
+// TODO(uladbohdan): to refactor, to extend.
+
+// TODO(uladbohdan): to figure out nesting in errors object.
+
+const createArticleValidator = ({ body }, res, next) => {
+  const errors = {};
+  if (body.locales) {
+    Object.keys(body.locales).forEach(loc => {
```
Comment (Drapegnik) [inline src/utils/validation.js]: use [`Object.entries`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/entries)

File: src/utils/validation.js [hunk 19]
```diff
@@ -1,8 +1,64 @@
...
+      ['title', 'subtitle', 'slug', 'text', 'locale'].forEach(field => {
+        if (!body.locales[loc][field]) {
+          errors[field] = 'must be presented';
+        }
+      });
+      if (loc !== body.locales[loc].locale) {
```
Comment (Drapegnik) [inline src/utils/validation.js]: how can this be possible?

File: src/utils/validation.js [hunk 20]
```diff
@@ -1,8 +1,64 @@
...
+        errors.localeConsistency = `bad locale structuring: ${loc} vs. ${body.locales[loc].locale}`;
+      }
+    });
+  }
+  if (!isEmpty(errors)) {
+    return next(new ValidationError(errors));
+  }
+  return next();
```
Comment (Drapegnik) [inline src/utils/validation.js]: does it works?

```js
return next(!isEmpty(errors) && new ValidationError(errors));
```

File: src/utils/validation.js [hunk 21]
```diff
@@ -22,10 +78,18 @@ export const checkIsFound = object => {
   if (!object) {
     throw new HttpError(404);
   }
+  return object;
+};
 
+export const checkDiaryIsFound = object => {
```
Comment (Drapegnik) [inline src/utils/validation.js]: redundant, you can achieve this with:
```js
const checkIsFound = (object, code = 404) ...

// for dairy:
.then(d => checkIsFound(d, 204))
```

File: src/utils/validation.js (hunk 20 above)
Comment (UladBohdan) [reply src/utils/validation.js]: it works fine, yes.

File: src/utils/validation.js (hunk 19 above)
Comment (UladBohdan) [reply src/utils/validation.js]: well something might go wrong in Frontend. It's better to check

File: src/utils/validation.js (hunk 18 above)
Comment (UladBohdan) [reply src/utils/validation.js]: done.

File: src/api/specials/diary/controller.js (hunk 15 above)
Comment (UladBohdan) [reply src/api/specials/diary/controller.js]: done.

File: src/api/article/localized/controller.js (hunk 14 above)
Comment (UladBohdan) [reply src/api/article/localized/controller.js]: I actually thought that is the plan for such cases.

File: src/api/user/model.js (hunk 17 above)
Comment (UladBohdan) [reply src/api/user/model.js]: not in one line, but looks nice. Done.

File: src/api/user/model.js (hunk 16 above)
Comment (UladBohdan) [reply src/api/user/model.js]: done.

File: src/api/article/controller.js (hunk 13 above)
Comment (UladBohdan) [reply src/api/article/controller.js]: How difference is worse?

File: src/api/article/controller.js (hunk 12 above)
Comment (UladBohdan) [reply src/api/article/controller.js]: No reassignment is for linter to shut up.

File: src/api/article/controller.js (hunk 10 above)
Comment (UladBohdan) [reply src/api/article/controller.js]: done.

File: src/api/article/controller.js (hunk 9 above)
Comment (UladBohdan) [reply src/api/article/controller.js]: it was indeed :(

File: src/api/article/controller.js (hunk 12 above)
Comment (Drapegnik) [reply src/api/article/controller.js]: in such cases better use `ignore`

File: src/api/article/controller.js (hunk 3 above)
Comment (UladBohdan) [reply src/api/article/controller.js]: done.

File: src/api/article/controller.js (hunk 13 above)
Comment (Drapegnik) [reply src/api/article/controller.js]: for me `filter` is more clear and understandable

**Comment by UladBohdan** [comment]
[to return best matches for prev/next diaries rather than Boolean](https://trello.com/c/pSOSJmbU/175-to-return-best-matches-for-prev-next-diaries-rather-than-boolean)

**Pull request merged**